### PR TITLE
make remat reduce precision of saved values to avoid xla excess precision

### PR DESCRIPTION
**Pull request by mattjj**
tl;dr Fix a `jax.remat`+XLA interaction that could cause subtle numerical issues in gradients due to values being saved for the backward pass in a different precision than how they're represented on the forward pass. The fix is to introduce `jax.lax.reduce_precision` calls in the forward pass to ensure XLA's excess precision isn't used on forward-pass consumers of those values.

An implication of this change is that it may cause `f(x) != jax.value_and_grad(f)(x)[0]` (due to numerical differences) for more functions `f` than is currently the case. The reason is that if `f` contains `jax.remat` calls, then `lambda x: jax.value_and_grad(f)(x)[0]` may contain some `jax.lax.reduce_precision` calls that are not in `f`. However, we think this is not problematic because for `jax.jit`-decorated functions we likely already have `f(x) != jax.value_and_grad(f)(x)[0]` simply due to the compiler optimizing the two programs differently.

---

XLA has a feature called ["excess precision"](https://github.com/openxla/xla/blob/3a050122daa3d930c0064396819a8186b4a0db49/xla/xla.proto#L180-L186), which can be turned off but defaults to on:
```
  // Allows xla to increase the output precision of floating point operations
  // and all floating-point conversions to be simplified, including those
  // that affect the numerics. The `FloatNormalization` pass inserts many
  // `f32 -> bf16 -> f32` conversion pairs. These are not removed by the
  // `AlgebraicSimplifier`, as that will only simplify conversions that are
  // no-ops, e.g. `bf16 -> f32 -> bf16`. Removing these improves accuracy.
```

Ah yes, "improves accuracy". The path to numerical hell is paved with "improves accuracy". (Actually, I'm exaggerating: I don't think I've heard any complaints about this XLA feature until we found this remat interaction.)

Here's an example of excess precision in action on TPU, with no `jax.remat` involved yet:

```python
import jax
import jax.numpy as jnp

def f(x):
  return jnp.dot(x, x) * jnp.float32(1.)

finfo = jnp.finfo('bfloat16')
x = (jnp.bfloat16(1) + finfo.eps) * jnp.eye(2, dtype='bfloat16')
print(f(x))
print(jax.jit(f)(x))
```

```
Array([[1.015625, 0.      ],
       [0.      , 1.015625]], dtype=float32)
Array([[1.015686, 0.      ],
       [0.      , 1.015686]], dtype=float32)
```

Different results for `f(x)` and `jax.jit(f)(x)`! In fact, the `jax.jit(f)(x)` results look like we computed things in a higher precision, namely float32. What happened here is that XLA saw this input HLO program:

```
HloModule jit_f, entry_computation_layout={(bf16[2,2]{1,0})->f32[2,2]{1,0}}

ENTRY main.4 {
  Arg_0.1 = bf16[2,2]{1,0} parameter(0)
  dot.2 = bf16[2,2]{1,0} dot(Arg_0.1, Arg_0.1), lhs_contracting_dims={1}, rhs_contracting_dims={0}
  ROOT convert.3 = f32[2,2]{1,0} convert(dot.2)
}
```

That is, we compute a (bf16, bf16) -> bf16 dot, and then consume its result in a float32 computation (in this case just simplified to a dtype conversion to f32). The "excess precision" logic in XLA says: "Hey, on TPU I accumulated the result of the dot in f32 anyway; why would I cast it down to bf16 when the consumer just converts back to f32 again? Instead, I'll keep it in f32 and avoid the downcast-then-upcast. Improves accuracy, and faster too!"

So far, so good. Sure, it means `f` and `jit(f)` differ, but is anyone going to complain about getting better precision results?

Well, `jax.remat` is. The issue arises when we save the bf16 dot result for the backward pass: the downcast is elided on the forward pass for the reason above (XLA sees the consumer converts to f32 and skips the downcast), but not on the backward pass because `jax.remat` doesn't let XLA optimize between the forward and backward passes (to prevent CSE).

Here is a very contrived program showing the issue with `jax.remat`:

```python
import jax
import jax.numpy as jnp

finfo = jnp.finfo('bfloat16')
eps = finfo.eps

@jax.custom_vjp
def dot(x):
  return jnp.dot(x, x)
def dot_fwd(x):
  return dot(x), None
def dot_bwd(_, g):
  return g,
dot.defvjp(dot_fwd, dot_bwd)

@jax.custom_vjp
def foo(x):
  return jnp.float32(1.) * x
def foo_fwd(x):
  return foo(x), x
def foo_bwd(x, _):
  return jnp.float32(1.) * x,
foo.defvjp(foo_fwd, foo_bwd)

@jax.jit
@partial(jax.remat, policy=lambda *_, **__: True)
def f(x):
  x = dot(x)
  return foo(x)

x = (jnp.bfloat16(1) + eps) * jnp.eye(2, dtype='bfloat16')
y, vjp = jax.vjp(f, x)
y_, = vjp(jnp.ones_like(y))

print(y)
print(y_)
```

```
Array([[1.015686, 0.      ],
       [0.      , 1.015686]], dtype=float32)
Array([[1.015625, 0.      ],
       [0.      , 1.015625]], dtype=float32)
```

It's basically the same issue, but now even though everything is under a `jax.jit`, we're seeing a different value in the forward and backward passes due to excess precision only on the forward pass.

This behavior can cause bad and difficult-to-debug issues with things like MoE models!

---

The fix in this PR is just to introduce explicit `jax.lax.reduce_precision` calls in the forward pass to disable XLA's excess precision for values that are saved.

An implication of this change is it's another reason why we may have `f(x) != jax.value_and_grad(f)(x)[0]`: if `f` contains `jax.remat` calls then the latter may now disable excess precision in cases where the former does not, changing numerics. But we think that is okay because we probably already have `f(x) != jax.value_and_grad(f)(x)[0]`, just due to XLA making different optimization decisions for the two programs in general.

### CODE REVIEW

File: jax/_src/ad_checkpoint.py [hunk 1]
```diff
@@ -544,10 +545,13 @@ def remat_partial_eval(trace, *tracers, jaxpr, **params):
   jaxpr_known, in_used_known = pe.dce_jaxpr(jaxpr_known, out_used_known)
   num_res = sum(used_res)
 
+  # insert reduce_precision calls
```
Comment (hawkinsp) [inline jax/_src/ad_checkpoint.py:548]: I'm going to ask for a slightly better comment, or at least a link to the excellent PR.

**Review by hawkinsp — APPROVED** [review]
